server/src/main/java/org/elasticsearch/cluster/metadata/ResettableValue.java

```diff
@@ -0,0 +1,216 @@
+     * responsibility of the caller to preserve order of the fields and their backwards compatibility.
+     *
+     * @throws IOException
+     */
+    static <T> void write(StreamOutput out, ResettableValue<T> value, Writeable.Writer<T> writer) throws IOException {
+        out.writeBoolean(value.isDefined);
+        if (value.isDefined) {
+            out.writeBoolean(value.shouldReset());
+            if (value.shouldReset() == false) {
+                writer.write(out, value.get());
+            }
+        }
+    }
+
+    /**
+     * Reads a single optional and explicitly nullable value. This method is in direct relation with the
+     * {@link #write(StreamOutput, ResettableValue, Writeable.Writer)} which writes the respective value. It's the
+     * responsibility of the caller to preserve order of the fields and their backwards compatibility.
+     *
+     * @throws IOException
+     */
+    @Nullable
+    static <T> ResettableValue<T> read(StreamInput in, Writeable.Reader<T> reader) throws IOException {
+        boolean isDefined = in.readBoolean();
+        if (isDefined == false) {
+            return ResettableValue.undefined();
+        }
+        boolean shouldReset = in.readBoolean();
+        if (shouldReset) {
+            return ResettableValue.reset();
+        }
+        T value = reader.read(in);
+        return ResettableValue.create(value);
+    }
+
+    /**
+     * Gets the value and applies the function {@param f} when the value is not null.
+     */
+    public <U> U applyAndGet(Function<? super T, ? extends U> f) {
+        if (isDefined() == false || shouldReset()) {
+            return null;
+        } else {
+            return f.apply(value);
+        }
+    }
+
+    public <U> ResettableValue<U> map(Function<? super T, ? extends U> mapper) {
+        Objects.requireNonNull(mapper);
+        if (isDefined == false) {
+            return ResettableValue.undefined();
+        }
+        if (shouldReset()) {
+            return reset();
+        }
+        return ResettableValue.create(mapper.apply(value));
+    }
+
+    /**
+     * Applies the {@param mergeFunction} to the current value while respecting the different states:
+     * - If the newer value is reset, it returns {@link #undefined()} since it resets any previous values
+     * - If the newer value is undefined, it returns <code>this</code>
+     * - If the current value is undefined, it returns {@param other}
+     * - otherwise, it merges the two values based on the provided function
+     */
+    public ResettableValue<T> merge(ResettableValue<T> other, Function<T, T> mergeFunction) {
+        if (other.shouldReset()) {
+            return undefined();
+        }
+        if (other.isDefined() == false) {
+            return this;
+        }
+        if (this.isDefined() == false || this.shouldReset()) {
+            return other;
+        }
+        return this.map(mergeFunction);
```

Comment:
Nit: I think this always does `ResettableValue.create(mergeFunction.apply(value))`, right? I think I'd find that easier to understand.

Reply:
I will change it, but I also applied a comment that the previous steps need to be in place, otherwise we might get an NPE.